### PR TITLE
Add deprecation lint for duplicated `macro_export`s

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rust/issues/35896#issuecomment-381370556

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @rust-lang/lang 

Needs check-only crater run before proceeding.
ping @rust-lang/infra

**Comment by Mark-Simulacrum** [comment]
Crater run commenced, check-only, ETA ~3 days.

File: src/librustc_resolve/resolve_imports.rs [hunk 1]
```diff
@@ -887,14 +887,24 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
         // Since import resolution is finished, globs will not define any more names.
         *module.globs.borrow_mut() = Vec::new();
 
+        let report_already_exported = |this: &mut Self, ident, span, prev_span| {
+            let msg = format!("a macro named `{}` has already been exported", ident);
+            this.session.struct_span_err(span, &msg)
```
Comment (nikomatsakis) [inline src/librustc_resolve/resolve_imports.rs]: I'm ok with a hard error, but it seems like there *is* a defined semantics, right? (Shadowing) We could always issue warnings, in other words.

**Review by nikomatsakis — APPROVED** [review]
r=me from the technical perspective. I'm not 100% sure about the policy decision, but I guess let's see what the crater run has to say.

**Comment by Mark-Simulacrum** [comment]
Crater results: http://cargobomb-reports.s3.amazonaws.com/pr-50143/index.html

**Comment by alexcrichton** [comment]
Legitimate regressions

* [probor 0.3.0](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/probor-0.3.0/log.txt)
* [nom 4.0.0-beta3](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/nom-4.0.0-beta3/log.txt)
* [lsio-0.1.18](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/lsio-0.1.18/log.txt)
* [fixed-hash-0.2.1](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/fixed-hash-0.2.1/log.txt)
* [eagre-asn1-0.2.0](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/eagre-asn1-0.2.0/log.txt)
* [collenchyma-nn-0.3.4](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/collenchyma-nn-0.3.4/log.txt)
* [adhesion-0.5.0](http://cargobomb-reports.s3.amazonaws.com/pr-50143/a31461256fceb1acc1180d88e5b5fa5d843f213f/reg/adhesion-0.5.0/log.txt)

The `lsio` and `fixed-hash` regressions seemed to transitively affect a lot of other dependencies.

Sounds like this may want to start out as a warning?

**Comment by durka** [comment]
What would be the motivation for a break here? The whole `#[macro_export]` system is going to be deprecated anyway, right?

**Comment by petrochenkov** [comment]
@nikomatsakis 
>it seems like there is a defined semantics, right? (Shadowing)

The exact problem is that there's no defined semantics.
This "shadowing" works over the whole crate, not a single module and macros are "shadowed" in whatever unspecified order in which the resolution/expansion algorithm processes the crate (https://github.com/rust-lang/rust/issues/35896#issuecomment-381370556).

@durka 
While the whole Macro 1.0 system is a large widely used part of the language that's going to be supported forever even if deprecated, the `macro_export` duplication is a corner case that's more on the "kill it with fire" territory due to undefined semantics.

Since there's breakage, I'd prefer to turn this into a deny-by-default lint so dependent crates are not affected and send PRs to crates with root regressions.

**Comment by durka** [comment]
I realize that it got accidentally "defined" to work the way it does, but since it does work and it's not unsound or anything, it seems unfortunate to do the breaking change.

**Comment by nikomatsakis** [comment]
@petrochenkov I would have expected the order to be DFS, since that is significant to other parts of macro expansion. I'm surprised that is complex to achieve, but I'm probably not thinking especially clearly about it. Do you have an example where a problem might arise, or just a bad feeling? I see you wrote this:

> The order is generally determined by depth-first search, but I suspect things can become more complex with macro expansion.

Can you point at the relevant code?

(This walk is occurring at the end of expansion.)

**Comment by nikomatsakis** [comment]
(At minimum, some kind of active deprecation definitely makes sense -- and perhaps a hard error in Edition 2018.)

**Comment by petrochenkov** [comment]
@nikomatsakis 
>Do you have an example where a problem might arise, or just a bad feeling?

Here, for example, `foo (1)` will "shadow" `foo (2)` for other crates due to expansion algorithm order.
At the same time, `foo (2)` will shadow `foo (1)` locally because it's defined later in the module.

```rust
macro_rules! generate_foo {
    () => {
        #[macro_export]
        macro_rules! foo { (1) => {} }
    }
}

generate_foo!();

#[macro_export]
macro_rules! foo { (2) => {} }

foo!(2); // Local use, OK
```

**Comment by petrochenkov** [comment]
I've replaced the error with a deprecation lint.

**Comment by nikomatsakis** [comment]
Nominating for brief discussion in the @rust-lang/lang meeting. I am inclined, I think, to go forward with a deprecation warning. I can definitely imagine that trying to support the current expansion order indefinitely will be quite difficult as we refactor things.

**Comment by Centril** [comment]
Forgot about checking nominations on issues / PRs on this repo; my bad!
Will have to discuss next meeting.

**Comment by nikomatsakis** [comment]
@Centril d'oh yeah we all forgot =) argh.

**Comment by cramertj** [comment]
@fcp close

We discussed this at the lang team meeting, and the consensus was that we should introduce a future-compatibility lint and follow up with an error after crates have a chance to update.

**Comment by joshtriplett** [comment]
@cramertj According to https://github.com/rust-lang/rust/pull/50143#issuecomment-385273233 , this proposal has been turned into precisely that compatibility lint.

**Comment by cramertj** [comment]
Oh, thanks. Might be good to rename the PR as to avoid confusion.

@rfcbot fcp merge

File: src/librustc/lint/builtin.rs [hunk 2]
```diff
@@ -273,6 +274,12 @@ declare_lint! {
     "detects name collision with an existing but unstable method"
 }
 
+declare_lint! {
+    pub DUPLICATED_MACRO_EXPORTS,
```
Comment (petrochenkov) [inline src/librustc/lint/builtin.rs]: I've just realized that "duplicate" may be more correct English here than "duplicated".

**Comment by petrochenkov** [comment]
Rebased.

**Pull request merged**